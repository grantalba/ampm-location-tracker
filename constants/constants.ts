export type OnboardingScreensType = {
  id: number;
  image: any;
  title: string;
  desc: string;
};

export const onboarding_screens: OnboardingScreensType[] = [
  {
    id: 1,
    image: require("@/assets/images/favicon.png"),
    title: "Track Your Location",
    desc: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. In dapibus varius nulla, quis dictum augue.",
  },
  {
    id: 2,
    image: require("@/assets/images/favicon.png"),
    title: "Search Location",
    desc: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. In dapibus varius nulla, quis dictum augue.",
  },
  {
    id: 3,
    image: require("@/assets/images/favicon.png"),
    title: "Light/Dark Theme",
    desc: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. In dapibus varius nulla, quis dictum augue.",
  },
];
