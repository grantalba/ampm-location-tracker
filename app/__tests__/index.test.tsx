import { useRouter } from "expo-router";
import React from "react";
import renderer from "react-test-renderer";

import { onboarding_screens } from "@/constants/constants";
import { NavigationContainer } from "@react-navigation/native";
import Index from "../index";

jest.useFakeTimers();

// Mock expo-router
jest.mock("expo-router", () => ({
  useRouter: jest.fn(),
}));

// Mock react-native-safe-area-context
jest.mock("react-native-safe-area-context", () => ({
  useSafeAreaInsets: () => ({
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  }),
}));

const mockNavigate = jest.fn();
beforeEach(() => {
  (useRouter as jest.Mock).mockReturnValue({ navigate: mockNavigate });
});

const renderWithProviders = () =>
  renderer.create(
    <NavigationContainer>
      <Index />
    </NavigationContainer>
  );

describe("Index (Onboarding) Screen with react-test-renderer", () => {
  it("should match the snapshot", () => {
    const tree = renderWithProviders();
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("renders first onboarding screen", async () => {
    const tree = renderWithProviders();

    const title = tree.root.findAll(
      (node) =>
        typeof node.type === "string" &&
        node.props.children === onboarding_screens[0].title
    );

    expect(title.length).toBeGreaterThan(0);

    const desc = tree.root.findAll(
      (node) =>
        typeof node.type === "string" &&
        node.props.children === onboarding_screens[0].desc
    );
    expect(desc.length).toBeGreaterThan(0);

    const nextButton = tree.root.findAll(
      (node) => typeof node.type === "string" && node.props.children === "Next"
    );
    expect(nextButton.length).toBeGreaterThan(0);
  });

  it("renders second onboarding screen", () => {
    const tree = renderWithProviders();
    const button = tree.root.find(
      (node) => node.props.accessibilityLabel === "onboardingTextButton"
    );
    button.props.onPress();

    const title = tree.root.find(
      (node) =>
        typeof node.type === "string" &&
        node.props.children === onboarding_screens[1].title
    );
    expect(title).toBeDefined();

    const desc = tree.root.findAll(
      (node) =>
        typeof node.type === "string" &&
        node.props.children === onboarding_screens[1].desc
    );
    expect(desc.length).toBeGreaterThan(0);
  });

  it("renders third onboarding screen and navigates on last press", () => {
    const tree = renderWithProviders();
    const button = tree.root.find(
      (node) => node.props.accessibilityLabel === "onboardingTextButton"
    );

    // simulate presses for each onboarding screen
    for (let i = 1; i <= onboarding_screens.length; i++) {
      button.props.onPress();
    }

    const finalTitle = tree.root.find(
      (node) =>
        typeof node.type === "string" &&
        node.props.children === onboarding_screens[2].title
    );
    expect(finalTitle).toBeDefined();

    const finalDesc = tree.root.findAll(
      (node) =>
        typeof node.type === "string" &&
        node.props.children === onboarding_screens[2].desc
    );
    expect(finalDesc.length).toBeGreaterThan(0);

    const letsGo = tree.root.find(
      (node) =>
        typeof node.type === "string" && node.props.children === "Let's Go"
    );
    expect(letsGo).toBeDefined();

    expect(mockNavigate).toHaveBeenCalledWith("/homescreen");
  });
});
