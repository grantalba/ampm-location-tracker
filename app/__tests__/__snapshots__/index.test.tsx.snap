// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index (Onboarding) Screen with react-test-renderer should match the snapshot 1`] = `null`;
